### ISSUE TITLE
fsleyes unable to open SPM12 .nii files having a JSON side-car file

### ISSUE THREAD

**Issue by halirutan** [pauldmccarthy/fsleyes#46]
Preface: This is not your fault!

When using SPM12 to convert Dicom files to NIFTI most of the meta-information is stored in side-car JSON files. However, these `.json` files are not [standard conform](https://www.json.org/json-en.html) because in JSON

> An array is an ordered collection of values. An array begins with `[` and ends with `]`. Values are separated by `,`.

Array values inside the JSON files exported by SPM12, however, have a trailing `,` which makes the Python `json` loader throw an error. Since fsleyes also uses the default JSON reader, it throws an error like the following

![fsl_error](https://user-images.githubusercontent.com/511683/77834153-c4196400-7142-11ea-8d40-3649187b7d97.png)

## Proposed solution

When opening a NIFTI, the side-car JSON is not required. In case the JSON loader throws an error, just neglect the meta-information and open the image.

If you like, I can provide a PR for this.

**Comment by pauldmccarthy** [comment]
Thanks for the report; even though the JSON file is poorly formed, it definitely should not cause a load error - at most a warning should be emitted.

The relevant code is actually in [fslpy](https://github.com/pauldmccarthy/fslpy/blob/630b9e1a3f4dd00cdb79b140d3ad4b5e7f04fcd5/fsl/data/image.py#L1168). I'm about to release a new version of fslpy so will sneak in a fix for this.

**Comment by halirutan** [comment]
> will sneak in a fix for this

That would be great.

I do believe that the Python JSON parser does the right thing by emitting an error. Especially since the JSON is a simple format with a clearly defined standard and programs should follow that. However, there is a way to fix the Python JSON loading by defining a custom JSON decoder. Basically, it's Copy&Paste with small adjustments where the error is raised. So if you do want to use the meta-information, you can use something like the code at the end.

The only difference is the `try` part in the while loop, where I append a value only if there is no error and I ignore `StopIteration` exceptions.

```
"""Hack for reading the supposed-to-be JSON files that SMP12 produces"""
import json
import re
from json.decoder import JSONDecodeError


def load(fp):
    return json.load(fp, cls=FuckSON)


class FuckSON(json.JSONDecoder):
    def __init__(self, *, object_hook=None, parse_float=None, parse_int=None, parse_constant=None, strict=True,
                 object_pairs_hook=None):
        super().__init__(object_hook=object_hook, parse_float=parse_float, parse_int=parse_int,
                         parse_constant=parse_constant, strict=strict, object_pairs_hook=object_pairs_hook)
        self.parse_array = fuckSONArray
        self.scan_once = json.scanner.py_make_scanner(self)


FLAGS = re.VERBOSE | re.MULTILINE | re.DOTALL
WHITESPACE = re.compile(r'[ \t\n\r]*', FLAGS)
WHITESPACE_STR = ' \t\n\r'


def fuckSONArray(s_and_end, scan_once, _w=WHITESPACE.match, _ws=WHITESPACE_STR):
    s, end = s_and_end
    values = []
    nextchar = s[end:end + 1]
    if nextchar in _ws:
        end = _w(s, end + 1).end()
        nextchar = s[end:end + 1]
    # Look-ahead for trivial empty array
    if nextchar == ']':
        return values, end + 1
    _append = values.append
    while True:
        try:
            value, end = scan_once(s, end)
            _append(value)
        except StopIteration as err:
            print("Damn it")
        nextchar = s[end:end + 1]
        if nextchar in _ws:
            end = _w(s, end + 1).end()
            nextchar = s[end:end + 1]
        end += 1
        if nextchar == ']':
            break
        elif nextchar != ',':
            raise JSONDecodeError("Expecting ',' delimiter", s, end - 1)
        try:
            if s[end] in _ws:
                end += 1
                if s[end] in _ws:
                    end = _w(s, end + 1).end()
        except IndexError:
            pass

    return values, end
```

**Comment by pauldmccarthy** [comment]
I released a new version of fslpy yesterday, so if you are running FSLeyes from a python environment, you should be able to update fslpy to prevent the error. If you are using a standalone version of FSLeyes, you will have to wait a little while, as I won't be able to create new builds for at least a few days.

I think that in this instance the SPM developers would be happy to adjust their JSON export code to conform to the JSON standard. In fact (without running the code), it looks to me like this should be handled: 

https://github.com/spm/spm12/blob/3085dac00ac804adb190a7e82c6ef11866c8af02/spm_jsonwrite.m#L167

You're definitely using the most recent version of SPM12? If so (and it is a bug in SPM), I'm happy to send an email to their bug report address.

**Comment by halirutan** [comment]
> You're definitely using the most recent version of SPM12?

I was staring at the exact same line you linked for a long time :) but it's not the right one. Numerical vectors [are exported differently](https://github.com/spm/spm12/blame/3085dac00ac804adb190a7e82c6ef11866c8af02/spm_jsonwrite.m#L220-L221).

Since I didn't export the data, I tried to trace, why we're not at the last version that includes the fix. The header of said JSON files says (note that in exported cells below have no trailing comma):

```
"version": [
    "matlab 9.3.0.713579 (R2017b)",
    "spm_dicom_convert.m - SPM12 (7219)"
]
```

Long story short, the bug is already fixed in SPM r7486 but for some reason, an old version was used for exporting and even our compute-servers only have r7486.

Thanks for your help! Now we don't have to re-export all legacy datasets.

**Comment by pauldmccarthy** [comment]
I think this problem has been resolved - feel free to re-open if you are still having problems.
